rust/bridge/shared/src/node/convert.rs

```diff
@@ -632,47 +631,170 @@ pub(crate) unsafe fn extend_lifetime<'a, 'b: 'a, T: ?Sized>(some_ref: &'a T) ->
 /// The name of the property on JavaScript objects that wrap a boxed Rust value.
 pub(crate) const NATIVE_HANDLE_PROPERTY: &str = "_nativeHandle";
 
+/// A marker trait for Rust objects exposed to JavaScript through [`neon::types::JsBox`].
+///
+/// Since this trait is used as a marker and implemented through a macro,
+/// operations that might differ across types have been factored into the [`BridgeHandleStrategy`]
+/// trait. The [`bridge_handle`] macro will automatically choose the correct strategy.
+pub trait BridgeHandle: Send + Sync + Sized + 'static {
+    /// Factors out operations that differ between mutable and immutable bridge handles.
+    type Strategy: BridgeHandleStrategy<Self>;
+}
+
+/// Factors out operations that differ between mutable and immutable bridge handles.
+///
+/// See the [`BridgeHandle`] trait.
+pub trait BridgeHandleStrategy<T> {
+    /// If the Rust object needs any additional wrapping before being boxed by Neon,
+    /// this type can be used.
+    type JsBoxContents: From<T> + Send + 'static;
+}
+
+/// A phantom type for immutable bridge handles.
+///
+/// See [`BridgeHandleStrategy<T>`].
+pub struct Immutable<T>(std::marker::PhantomData<T>);
+
+/// A phantom type for mutable bridge handles.
+///
+/// See [`BridgeHandleStrategy<T>`].
+pub struct Mutable<T>(std::marker::PhantomData<T>);
+
+/// Immutable handles need no extra boxing.
+impl<T: BridgeHandle> BridgeHandleStrategy<T> for Immutable<T> {
+    type JsBoxContents = T;
+}
+
+/// Mutable handles use a [`RefCell`] for safety.
+impl<T: BridgeHandle> BridgeHandleStrategy<T> for Mutable<T> {
+    type JsBoxContents = RefCell<T>;
+}
+
+/// Shorthand for `T::Strategy::JsBoxContents` when `T: `[`BridgeHandle`], which Rust can't always
+/// resolve.
+type BoxContentsFor<T> = <<T as BridgeHandle>::Strategy as BridgeHandleStrategy<T>>::JsBoxContents;
```

Comment:
Some concrete examples:

```
BoxContentsFor<PublicKey> = PublicKey
BoxContentsFor<HSMEnclaveClient> = RefCell<HSMEnclaveClient>
```

"Of" doesn't sound as good to me because that feels like it's drilling into the type instead of wrapping it in something. But ultimately this is bikeshedding and whatever makes more sense to you is probably best.